File: hackerrank/project_euler/problem_4/largest_palindrome_product.js
// https://www.hackerrank.com/contests/projecteuler/challenges/euler004/problem

const larPal = n => {
    let largest = Number.MIN_SAFE_INTEGER;
    for (let i = Math.floor(n ** 0.5) + 1; i > 99; i--) {
        for (let j = Math.floor(n ** 0.5) - 10; j < 1000; j++) {
            let product = i * j;
            if (product < n && isPal(product)) {
                largest = largest < product ? product : largest;
            }
        }
    }
    return largest;
};

const isPal = n => n == n.toString().split("").reverse().join("");

console.log(larPal(101110), " === ", 101101);
console.log(larPal(101101), " === ", 99999);
console.log(larPal(800000), " === ", 793397);

File: hackerrank/project_euler/problem_4/largest_palindrome_product.py
# https://www.hackerrank.com/contests/projecteuler/challenges/euler004/problem

def isPal(n):
    n = str(n)
    return n == n[::-1]


def larPal(n):
    pals = []
    for i in range(99, int(n**0.5)+10):
        for j in range(999, int(n**0.5)-10, -1):
            product = i*j
            if isPal(product) and product < n:
                pals.append(product)
    return sorted(pals)[-1]


print(larPal(101111))
